tests/dsp_testing/filters.ipynb

```diff
@@ -0,0 +1,947 @@
+    "    # with the first transmitted pulse. This is done automatically when Borealis runs.\n",
+    "    extra_samples = 0\n",
+    "    for dm, stage_taps in zip(reversed(scheme.dm_rates), reversed(filter_taps)):\n",
+    "        extra_samples = (extra_samples * dm) + len(stage_taps) // 2\n",
+    "    sample_diff = default_extra_samples - extra_samples\n",
+    "    if sample_diff >= 0:\n",
+    "        input_samples = input_samples[:, sample_diff:]\n",
+    "    else: \n",
+    "        print((input_samples.shape[0], input_samples.shape[1] + np.abs(sample_diff)))\n",
+    "        padded_input = np.zeros((input_samples.shape[0], input_samples.shape[1] + np.abs(sample_diff)), dtype=input_samples.dtype)\n",
+    "        padded_input[:, -sample_diff:] = input_samples\n",
+    "        input_samples = padded_input\n",
+    "        \n",
+    "    # Move samples to GPU if cupy is being used.\n",
+    "    if cupy_available:\n",
+    "        input_samples = cp.asarray(input_samples)\n",
+    "\n",
+    "    # Now actually apply the filter to the data\n",
+    "    processor.apply_filters(input_samples)\n",
+    "\n",
+    "    filter_outputs = processor.filter_outputs\n",
+    "    mixed_filters = processor.filters\n",
+    "    \n",
+    "    # Move the results to the CPU (if GPU is used, this is needed)\n",
+    "    if cupy_available:\n",
+    "        cpu_outputs = []\n",
+    "        mixed_filters_cpu = []\n",
+    "        for i in range(len(filter_outputs)):\n",
+    "            cpu_outputs.append(cp.asnumpy(filter_outputs[i]))\n",
+    "            mixed_filters_cpu.append(cp.asnumpy(mixed_filters[i]))\n",
+    "        filter_outputs = cpu_outputs\n",
+    "        mixed_filters = mixed_filters_cpu\n",
+    "\n",
+    "    del processor\n",
+    "        \n",
+    "    return filter_outputs, mixed_filters\n",
+    "\n",
+    "def plot_dsp_process(scheme, input_samples, output_samples, filter_taps, mixing_freqs, zoom_in=False, freq_lim=None):\n",
+    "    for i in range(len(scheme.stages)):\n",
+    "        antenna_output = output_samples[i][0, 0, :]\n",
+    "        filter = filter_taps[i]\n",
+    "        \n",
+    "        if zoom_in:\n",
+    "            freq_lim = np.array([-scheme.stages[i].output_rate, scheme.stages[i].output_rate])\n",
+    "        else:\n",
+    "            freq_lim = freq_lim\n",
+    "            \n",
+    "        if i == 0:\n",
+    "            if zoom_in:\n",
+    "                freq_lim -= mixing_freqs[0]\n",
+    "            title = f'Filtered Rawrf Samples - Stage {i}'\n",
+    "            plot_filter_stage(input_samples, antenna_output, filter[0, :],\n",
+    "                              scheme.stages[i].input_rate, scheme.stages[i].output_rate,\n",
+    "                              title, center_freq=mixing_freqs[0], freq_lim=freq_lim)\n",
+    "        else:\n",
```

Comment:
The stage 0 filter looks offset in the time domain. Is the start of the filter aligned to the transmit pulse or to the TR signal?

![image](https://github.com/SuperDARNCanada/borealis/assets/25570527/d73428e0-893b-4095-8b57-d6eefaab98bf)

Reply:
The filter is offset, as any output samples affected by filter edge effects are thrown out. This is how the first pulse eventually aligns to the zeroth sample.